src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs

```diff
@@ -165,8 +182,9 @@ private static ValueGetter<VBuffer<ReadOnlyMemory<char>>> GetKeyValueGetter<TKey
             var keyType = GetPrimitiveType(typeof(TKey), out bool isKeyVectorType);
             var valueType = GetPrimitiveType(typeof(TValue), out bool isValueVectorType);
             var dataViewBuilder = new ArrayDataViewBuilder(env);
-            dataViewBuilder.AddColumn(keyColumnName, keyType, keys.ToArray());
```

Comment:
>keyType [](start = 53, length = 7)

in old code if you pass TKey as string, keyType for it would be ReadOnlyMemory<char>, why you need what wrapping around?

Reply:
Unfortunately the old code did not work if TKey was of string type. 

So here is what happens:
ArrayDataViewBuilder.AddColumn<T> will fail if T is of type string. The reason for this is for two reasons:
1) The PrimtiveType.RawType is a ReadonlyMemory<char> which does not match type of string
2) The string values are not converted to ReadOnlyMemory. 

ArrayDataViewBuilder.AddColumn does have an implementation that takes in a string, but that is not called when you are calling it from a function that is using a generic argument. So I added this wrapping that will check if the type is string, if so, convert to ReadOnlyMemoryChar and pass that to the builder.


---
In reply to: [246977957](https://github.com/dotnet/machinelearning/pull/2115#discussion_r246977957) [](ancestors = 246977957)